components/dada-ir/src/lines.rs

```diff
@@ -1,70 +1,144 @@
+        self.lines[line0].start
     }
 
     fn num_lines(&self) -> usize {
-        self.line_endings.len() + 1
+        self.lines.len()
+    }
+
+    fn offset(&self, position: LineColumn) -> Offset {
+        if position.line0_usize() >= self.num_lines() {
+            return self.end_offset;
+        }
+        let line_start = self.line_start(position.line0_usize());
+        (line_start + position.column0()).min(self.end_offset)
+    }
+
+    fn line_column(&self, position: Offset) -> LineColumn {
+        match self.lines.binary_search_by_key(&position, |l| l.start) {
+            Ok(line0) => LineColumn::new0(line0, 0u32),
+            Err(next_line0) => {
+                let line0 = next_line0 - 1;
+                let line = &self.lines[line0];
+                // not quite column yet, because there may be wide characters between line start and position
+                // at this point it's the byte offset from line start
+                // we need to adjust for it
+                let mut column0 = position - line.start;
+                for wc in line.wide_chars.iter() {
+                    if wc.start >= position {
+                        break;
+                    }
+                    // e.g.: 🙂 will have len 4, but we count it as 1 character, so we substract 3
+                    column0 -= wc.len() - 1;
+                }
+                LineColumn::new0(line0, column0)
+            }
+        }
     }
 }
 
 /// Converts a character index `position` into a line and column tuple.
 pub fn line_column(db: &dyn crate::Db, filename: Filename, position: Offset) -> LineColumn {
     let table = line_table(db, filename);
-    match table.line_endings.binary_search(&position) {
-        Ok(line0) | Err(line0) => {
-            let line_start = table.line_start(line0);
-            LineColumn::new0(line0, position - line_start)
-        }
-    }
+    table.line_column(position)
 }
 
 /// Given a (1-based) line/column tuple, returns a character index.
 pub fn offset(db: &dyn crate::Db, filename: Filename, position: LineColumn) -> Offset {
     let table = line_table(db, filename);
-
-    if position.line0_usize() >= table.num_lines() {
-        return table.end_offset;
-    }
-    let line_start = table.line_start(position.line0_usize());
-    (line_start + position.column0()).min(table.end_offset)
+    table.offset(position)
 }
 
 #[salsa::memoized(in crate::Jar ref)]
 #[allow(clippy::needless_lifetimes)]
 fn line_table(db: &dyn crate::Db, filename: Filename) -> LineTable {
     let source_text = crate::manifest::source_text(db, filename);
-    let mut p: usize = 0;
-    let mut table = LineTable {
-        line_endings: vec![],
-        end_offset: Offset::from(source_text.len()),
-    };
-    for line in source_text.lines() {
-        p += line.len();
-        table.line_endings.push(Offset::from(p));
-        p += 1;
+    LineTable::new(source_text)
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Maybe we should use our own test harness, like https://github.com/dada-lang/dada/blob/main/dada_tests/heap-graph/cursor-position.dada ?  But I think the tests you  wrote here are more intuitive.

Reply:
The way I see it this are unit tests and the ones from the test harness are integration tests, so I think they are complementary.

Reply:
Agreed, this seems like a classic and perfect use for unit tests